index.d.ts

```diff
@@ -0,0 +1,19 @@
+type Points = number[][]
+
+declare class KDBush<T> {
+  ids: number[]
+  coords: number[]
+  nodeSize: number
+  points: T[]
+  range(minX: number, minY: number, maxX: number, maxY: number)
+  within(x: number, y: number, r: number)
+}
+
+interface KDBush<T> {
+  (points: Points): KDBush<Points>
+  <T>(points: T[], getX: (p: T) => void, getY: (p: T) => void, nodeSize?: number, ArrayType?: any): KDBush<T>
+}
+
+declare const kdbush: KDBush<Points>
```

Comment:
lol... I know... it's a bit intense, ok here goes:

There's two different types of inputs (assumed & custom)

## Assumed
This one is pretty straight forward and is automatically assigned by default
```typescript
type Points = number[][]
```
## Custom
The mythical `<T>` (can be defined as any letter or word), this mutates based on the user's input and that same type can be translated to other properties/methods/class.

```typescript
interface KDBushStatic {
  (points: Points): KDBush<Points>
  <T>(points: T[], getX: (p: T) => void, getY: (p: T) => void, nodeSize?: number, ArrayType?: any): KDBush<T>
}
```
Lets look at this code above:
- User inputs custom points (not assumed type), then he is **required** to include `getX()` & `getY()` callbacks (types will fail if not provided).
- When the user inputs his custom Array of points, that same input type is then translated to `getX()` & `getY()` which enable the type checking based on custom entry, which makes it easier to define `X` & `Y` values.

```typescript
declare class KDBush<T> {
  points: T[]
}
```
- The KDBush Class also has a `<T>` which allows you to translate the types given by the Array of `points` from the main method. This allows you to define the type of Array of `points` inside the class.

@mourner Let me know if this helps in any way, it's definitely a bit confusing a first (this only happens when the user is able to define his own input types).

## Changes

Few minor changes (next commit), renamed the `interface` to `KDBushStatic` (no longer requires the `<T>`)

```typescript
interface KDBushStatic {
  (points: Points): KDBush<Points>
  <T>(points: T[], getX: (p: T) => void, getY: (p: T) => void, nodeSize?: number, ArrayType?: any): KDBush<T>
}

declare const kdbush: KDBushStatic
declare namespace kdbush {}
export = kdbush
```